cpp/src/neighbors/mg/snmg.cuh

```diff
@@ -92,13 +92,28 @@ void build(const raft::resources& clique,
     RAFT_LOG_DEBUG("REPLICATED BUILD: %d*%drows", index.num_ranks_, n_rows);
 
     index.ann_interfaces_.resize(index.num_ranks_);
-#pragma omp parallel for
+
+    // Enable nested parallelism
+    int saved_omp_threads = cuvs::core::omp::get_max_threads();
+    int threads_per_rank  = std::max(1, saved_omp_threads / index.num_ranks_);
+    cuvs::core::omp::set_nested(1);
+
+    const int& requirements = index.num_ranks_;
+    cuvs::core::omp::check_threads(requirements);
+
+#pragma omp parallel for num_threads(index.num_ranks_)
     for (int rank = 0; rank < index.num_ranks_; rank++) {
+      // Set thread limit for this rank's nested OpenMP regions
+      cuvs::core::omp::set_num_threads(threads_per_rank);
+
       const raft::resources& dev_res = raft::resource::set_current_device_to_rank(clique, rank);
       auto& ann_if                   = index.ann_interfaces_[rank];
       cuvs::neighbors::build(dev_res, ann_if, index_params, index_dataset);
       resource::sync_stream(dev_res);
     }
+
+    // Restore original thread count
+    cuvs::core::omp::set_num_threads(saved_omp_threads);
```

Comment:
this restoration is a bit confusing. the pool of threads is set inside the for loop above, but its restored after the for loop? Can you verify if this logic is correct?